### ISSUE TITLE
accessibility: screen reader fails to read out tab names

### ISSUE THREAD

**Issue by phloxic** [flowplayer/flash#109]
According to http://flowplayer.org/forum/#!/moot/12676 

---

When I hover over the play/pause button JAWS doesn't read out Play/Pause. Also when I hit Tab button to tab over the various controls, JAWS keeps saying Tab instead of the title of the button (eg. Play, Pause, Enter Fullscreen etc).

---

**Comment by phloxic** [comment]
Minimal sample: http://flowplayer.blacktrash.org/test/accessibility.html

**Comment by danrossi** [comment]
We've gone over this in a previous ticket, jaws will say tab because you are tabbing and then read the accessibility name which is not the label but the button name, perhaps its a setting in Jaws to tell it not to say tab. I've changed it to use the user configured label instead if set to avoid the confusion what is uses and is in the current pull request

https://github.com/flowplayer/flash/blob/danielr-httpstreaming/lib/common/src/actionscript/org/flowplayer/ui/controllers/AbstractButtonController.as

I spent 2 hours getting it setup again to test and is working for me the last time i tried it out.

**Comment by danrossi** [comment]
Hi if you understand how the screen reader works, it works by tabbing not by using the mouse or mouse hovers. Upon each tab it will read its name or in the case of flash widgets it's accessibility name. Perhaps read the documentation of how Jaws works ? If someone has visibility issues they won't be using a mouse.

**Comment by phloxic** [comment]
I have now looked at this sample again - http://flowplayer.blacktrash.org/test/accessibility.html - and:
- you cannot get "into the player" by tabbing, you have to use the mouse to click the player first
- setting `SeamlessTabbing: "true"` does not change the behaviour - at least when it comes to tabbing I get the same behaviour when I omit this. Is this for Screen readers or IE only?

A user confirms my observations and claims it only works in IE - which I cannot test.

**Comment by danrossi** [comment]
jaws only works on IE

**Comment by danrossi** [comment]
or this feature does anyway

**Comment by phloxic** [comment]
And `SeamlessTabbing` _does_ make a difference in Jaws?

**Comment by phloxic** [comment]
Probably all the requirements only make a difference in Jaws and IE?
(using an A tag as container also make no difference in other circumstances)
Can you confirm?
Then this must be documented as well.

**Comment by danrossi** [comment]
It's an IE only feature I'm afraid therefore only tested on IE.

**Comment by phloxic** [comment]
IE only, or IE in combination with Jaws only?

**Comment by danrossi** [comment]
The JAWS feature only works with Flash on IE as there is the accesibility API for it. Therefore have only checked tabbing in IE which is needed for this.

**Comment by qualia100** [comment]
Our company uses flowplayer and we had the independent company to evaluate the player and had some critique about accessibility of the player (for us, it means 508 compliance).

The main point is that the player has to be accessed and controlled by keyboard only:

"By far the best situation is to ensure that the player's controls can be operated using standard keyboard commands for user interface controls - i.e. tabbing to buttons or button groups, and activating buttons using Enter or the spacebar; navigation within button groups that take up one tab stop should be supported by the arrow keys."

It is quite possible that JAWS are designed for IE only, but keyboard accessibility has to be working regardless of that. Some users might not needs JAWS but still have difficulty controlling a mouse.

**Comment by danrossi** [comment]
Anything other IE requires hacks see https://github.com/englandrp/Cross-browser-Flash-tabbing-solution

**Comment by danrossi** [comment]
not even happening with JWPlayer if you must setup a new ticket and label it as such as it's unrelated to the screen reader properties. It may require alot of work with the js api + hacks and changes in the player itself. That would have to get confirmed first though.

**Comment by qualia100** [comment]
I checked out that hack you suggested and I have to admit I am not that handy with actionscript. Has anyone tried it in a Flowplayer? It seems that there is a bit that has to be added to the flowplayer code.
And what exactly is not happening with JWPlayer - tabbing?

**Comment by danrossi** [comment]
This is a feature request yes

**Comment by qualia100** [comment]
Thanks for pointing me in the right direction - I was not able to use that solution, but it helped me to find my own. Now users can tab into the Flowplayer and out of it in most browsers (we are still testing...). However, the "tab" key goes through 5 'buttons' - Play Overlay, logo area, Play/Pause, Mute/Unmute and Fullscreen. What about the sliders - video duration and volume control? Why tab never accesses those fields - shouldn't users be able to access them as well?

**Comment by danrossi** [comment]
How are you giving flash focus ? The internal tabbing indexes have some things disabled because during these accessibilities updates it was for jaws therefore screen readers, you can't really control the sliders with the keyboard some key commands would have to be setup for that.

**Comment by qualia100** [comment]
I used the same idea as in this site:

http://freshmeathouse.channel4.com/#/hallway

More specifically, their javascript file that embeds the Flash object. When last html element before Flash object gets a focus, search for Flash object on a page and set focus on it. But if you try using tab on that site you will see that once you are in the Flash, you are stuck there. So to resolve that issue, I had to change AS code and that is a whole separate story. 

I just submitted my changes to include them in the next Flowplayer release, but I am not sure how this will work out - when I was making them I thought they will be just for our needs, not for everyone so they might not be able to include them.

**Comment by danrossi** [comment]
works !

**Comment by danrossi** [comment]
Going to close this and perhaps a feature request ticket is requiring for the focus / tabbing mod.
